### PR TITLE
Merge sender and receiver sc2sc validators

### PR DESCRIPTION
**Pull request by paologalligit**
Jira reference [ticket](https://horizenlabs.atlassian.net/browse/SDK-763)

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 1]
```diff
@@ -11,50 +11,59 @@ import com.horizen.certnative.BackwardTransfer
...
     with ModifierValidation[SidechainBlock]
     with SidechainTypes
     with NodeState
     with SparkzLogging
     with UtxoMerkleTreeView
-    with NetworkParamsUtils
-{
+    with NetworkParamsUtils {
 
+  protected val crossChainValidatorContainer: CrossChainValidatorContainer = new CrossChainValidatorContainer(
```
Comment (paolocappelletti) [inline sdk/src/main/scala/com/horizen/SidechainState.scala:60]: i would keep it simpler: we are defining a key-value structure but anyone wil be able to add more entries from outside the state
We can instead use a similar structure as AbstractSidechainNodeViewHolder.semanticBlockValidators :  

- define a trait Sc2ScBlockValidator similar to com.horizen.validation.SemanticBlockValidator with a method validate(block: PMOD)
-  here initialize a Seq of the above trait, with two implementations: one for output crosschainmessages and one for input

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 2]
```diff
@@ -181,18 +191,30 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
...
+      validate(tx, consensusEpochNumber, modWithdrawalEpochInfo.epoch).get
+      tx match {
+        case transaction: AbstractCrossChainRedeemTransaction =>
+          crossChainValidatorContainer.validateReceivingCrossChain(
+            new CrossChainRedeemMessageBodyToValidate(transaction)
+          )
+        case _ =>
+      }
+    })
+    crossChainValidatorContainer.validateSenderCrossChain(
```
Comment (paolocappelletti) [inline sdk/src/main/scala/com/horizen/SidechainState.scala:213]: with the above suggestion this piece becomes something like:
sc2scBlockValidators.foreach ( val.validate(mod))

File: sdk/src/main/scala/com/horizen/SidechainState.scala (hunk 1 above)
Comment (paologalligit) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:60]: That's a good point, gonna fix it

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 3]
```diff
@@ -11,52 +11,58 @@ import com.horizen.certnative.BackwardTransfer
...
     with UtxoMerkleTreeView
-    with NetworkParamsUtils
-{
-
-
+    with NetworkParamsUtils {
   override type NVCT = SidechainState
+  type PMOD = SidechainBlock
+
+  protected val crossChainValidators: Seq[CrossChainValidator[PMOD]] = Seq(
```
Comment (MarcoOl94) [inline sdk/src/main/scala/com/horizen/SidechainState.scala:62]: The idea is to pass the custom validators from the Application side. I think we need to add an additional injected object inside the **SidechainAppModule** like "Sc2ScMessageValidators" which is a `Seq[CrossChainValidator]` and pass it to the **SidechainState** constructor.

File: sdk/src/main/scala/com/horizen/validation/crosschain/sender/AbstractCrossChainMessageValidator.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import com.horizen.params.NetworkParams
+import com.horizen.sc2sc.{CrossChainMessageHash, Sc2ScConfigurator}
+import com.horizen.storage.SidechainStateStorage
+import com.horizen.transaction.MC2SCAggregatedTransaction
+import com.horizen.validation.crosschain.CrossChainValidator
+import com.horizen.{SidechainState, SidechainTypes}
+
+import scala.collection.JavaConverters._
+
+class AbstractCrossChainMessageValidator(
```
Comment (MarcoOl94) [inline sdk/src/main/scala/com/horizen/validation/crosschain/sender/AbstractCrossChainMessageValidator.scala:15]: This class is extended by the custom message validator from the application side. As we discussed we can add a **protected applicationValidate** that is called inside the validate and it is implemented by the custom validator.

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

**Review by paolocappelletti — APPROVED** [review]

**Pull request merged**